### DISCUSSION TITLE
What does pipe() mean?

### DISCUSSION THREAD

**Discussion by yyk808** [flomesh-io/pipy#180] [Q&A]
I mean `pipe()` not `pipeline()`.
To be more precise, what are the differences between pipe() and to()?

They both input a pipeline layout. And it seems like they both redirect( is this correct? ) the data to somewhere.  I got confused when reading code like `https://github.com/flomesh-io/ztm/blob/main/agent/mesh.js#L150-L151`.
As the question is not about `ztm`  itself,  I make a discussion here. 

Anyone can help me?

**Comment by pajama-coder** [comment]
@yyk808 Hi, it's a good question. Although the issue is closed, I still feel the need to clarify the confusing usage of `to()`. Unlike `pipe()`, an invocation of `to()` doesn't add any filters to the pipeline, instead, it should be looked at as a part of the filter that was just added prior to it. Filters that need to construct subpipelines all need a way for users to specify a subpipeline template, by calling `to()` is one way. The other is giving a direct function argument like we do for `pipe()`.

In short, `pipe()` is to create a `pipe` filter, but `to()` is for providing more info to the filter created before it.

**Reply by yyk808** [reply]
@pajama-coder Thank you for your clarification, now I got it.

However, there's still something I cannot understand completely about its usage. Here it is.
If I want to use pipe(), it must be in a declared pipeline.
```js
var pp = pipeline($=>$.dump())
pipeline($=>$.onStart(new Message('Msg')).pipe(pp)).spawn() // works fine.
pipy().task().onStart(new Message()).pipe(pp) // not work: not a function.
```
I can't directly call it from pipy(). Is this a intentional design?
